master/buildbot/reporters/irc.py

```diff
@@ -135,31 +321,80 @@ def kickedFrom(self, channel, kicker, message):
                                                            kicker,
                                                            message))
 
+    def userLeft(self, user, channel):
+        if user:
+            user = user.lower()
+        if channel:
+            channel = channel.lower()
+        if (channel, user) in self.contacts:
+            del self.contacts[(channel, user)]
+
+    def userKicked(self, kickee, channel, kicker, message):
+        self.userLeft(kickee, channel)
+
+    def userQuit(self, user, quitMessage=None):
+        if user:
+            user = user.lower()
+        for c, u in list(self.contacts):
+            if u == user:
+                del self.contacts[(c, u)]
+
+    results_colors = {
+        SUCCESS: 'GREEN',
+        WARNINGS: 'YELLOW',
+        FAILURE: 'RED',
+        EXCEPTION: 'PURPLE',
+        RETRY: 'AQUA_LIGHT',
+        CANCELLED: 'PINK',
+    }
+
+    short_results_descriptions = {
+        SUCCESS: ", Success",
+        WARNINGS: ", Warnings",
+        FAILURE: ", Failure",
+        EXCEPTION: ", Exception",
+        RETRY: ", Retry",
+        CANCELLED: ", Cancelled",
+    }
+
+    def format_build_status(self, build, short=False):
+        br = build['results']
+        if short:
+            text = self.short_results_descriptions[br]
+        else:
+            text = self.results_descriptions[br]
+        if self.useColors:
+            return "\x03{:d}{}\x0f".format(
+                _irc_colors.index(self.results_colors[br]),
+                text)
+        else:
+            return text
+
 
 class IrcStatusFactory(ThrottledClientFactory):
     protocol = IrcStatusBot
 
     shuttingDown = False
     p = None
 
-    def __init__(self, nickname, password, channels, pm_to_nicks, tags, notify_events,
+    def __init__(self, nickname, password, join_channels, pm_to_nicks, authz, tags, notify_events,
                  noticeOnChannel=False,
                  useRevisions=False, showBlameList=False,
                  parent=None,
-                 lostDelay=None, failedDelay=None, useColors=True, allowShutdown=False):
```

Comment:
I think we need a migration path for this feature as according to the Buildbot versioning polity we would have to do a major version bump. Perhaps it makes sense to print a deprecation warning when allowShutdown is passed and them modify authz to have similar effect as previous `allowShutdown` functionality.

Reply:
I consider this a security fix. However, adding decprecated params with a warning is easy, so what is a decision?

Reply:
We actually have three options:

1. Totally remove `allowForce` and `allowShutdown`. This is most secure and fully fixes #2957, however, it may break existing configurations.

2. Log a warning when `allowForce` or `allowShutdown` is used and **ignore it**. This way, the existing configurations will still work, but the bot behavior will change (with serious insecurity removed). This is an option I vote for now, until the next major version bump.

3. Log a warning and honor `allowForce` or `allowShutdown` by adding corresponding entries to `authz`. This way any insecure config will remain insecure (I guess there are not many od them, because of explicit warning agains  #2957 in the current documentation).

Reply:
@tardyp ^^

Reply:
Personally I think (1) is the best option, as silent breakage of Buildbot configuration is bad. Many people don't monitor logs for warnings.

Reply:
So after a discussion, I implemented option 3 in 3a83bf8.